File: tests/c_api/odc_header.c
/**
 * To build this program, please make sure to reference linked library:
 *
 *     gcc -lodccore -o odc-c-header odc_header.c
 */

#include <stdio.h>
#include <stdlib.h>

#include "odc/api/odc.h"

#define CHECK_RESULT(x) \
    do { \
        int rc = (x); \
        if (rc != ODC_SUCCESS) { \
            fprintf(stderr, "Error calling odc function \"%s\": %s\n", #x, odc_error_string(rc)); \
            exit(1); \
        } \
    } while (false); \

void usage() {
    fprintf(stderr, "Usage:\n");
    fprintf(stderr, "    odc-c-header <odb2 file 1> [<odb2 file 2> ...]\n\n");
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        usage();
        return 1;
    }

    // Initialise API
    CHECK_RESULT(odc_initialise_api());

    int argi;

    // Iterate over all supplied path arguments
    for (argi = 1; argi < argc; argi++) {
        char* path = argv[argi];

        odc_reader_t* reader = NULL;
        odc_frame_t* frame = NULL;

        // Open current path and initialise frame
        CHECK_RESULT(odc_open_path(&reader, path));
        CHECK_RESULT(odc_new_frame(&frame, reader));

        fprintf(stdout, "File: %s\n", path);

        int i = 0;
        int rc;

        // Iterate over all frames in the stream in non-aggregated mode, without decoding them
        while ((rc = odc_next_frame(frame)) == ODC_SUCCESS) {
            long row_count;
            int column_count;

            // Get row and column counts
            CHECK_RESULT(odc_frame_row_count(frame, &row_count));
            CHECK_RESULT(odc_frame_column_count(frame, &column_count));

            fprintf(stdout, "  Frame: %d, Row count: %ld, Column count: %d\n", ++i, row_count, column_count);

            int nproperties;

            // Get number of properties encoded in the frame
            CHECK_RESULT(odc_frame_properties_count(frame, &nproperties));

            const char* key;
            const char* value;

            int i;
            for (i = 0; i < nproperties; i++) {

                // Get property key and value by its index
                CHECK_RESULT(odc_frame_property_idx(frame, i, &key, &value));

                fprintf(stdout, "  Property: %s => %s\n", key, value);
            }

            int col;

            // Iterate over frame columns
            for (col = 0; col < column_count; ++col) {
                const char* name;
                int type;
                int element_size;
                int bitfield_count;

                // Get column information
                CHECK_RESULT(odc_frame_column_attributes(frame, col, &name, &type, &element_size, &bitfield_count));

                const char* type_name;

                // Lookup column type name
                CHECK_RESULT(odc_column_type_name(type, &type_name));

                fprintf(stdout, "    Column: %d, Name: %s, Type: %s, Size: %d\n", col + 1, name, type_name,
                        element_size);

                // Process bitfields only
                if (type == ODC_BITFIELD) {
                    int bf;

                    for (bf = 0; bf < bitfield_count; ++bf) {
                        const char* bf_name;
                        int bf_offset;
                        int bf_size;

                        // Get bitfield information
                        CHECK_RESULT(odc_frame_bitfield_attributes(frame, col, bf, &bf_name, &bf_offset, &bf_size));

                        fprintf(stdout, "      Bitfield: %d, Name: %s, Offset: %d, Nbits: %d\n", bf + 1, bf_name,
                                bf_offset, bf_size);
                    }
                }
            }

            fprintf(stdout, "\n");
        }

        // Unsuccessful end of frame iteration
        if (rc != ODC_ITERATION_COMPLETE) {
            fprintf(stderr, "Error: %s\n", odc_error_string(rc));
            return 1;
        }

        CHECK_RESULT(odc_free_frame(frame));

        // Deallocate reader instance
        CHECK_RESULT(odc_close(reader));
    }

    return 0;
}

File: rust/crates/odc/examples/odc_header.rs
//! Print information extracted from the header of each frame in ODB-2
//! files, without decoding the data.
//!
//! Usage: `cargo run --example odc_header -- <file.odb> [<file2.odb> ...]`

use odc::{ColumnType, Reader, ReaderOptions};

const fn type_name(column_type: ColumnType) -> &'static str {
    match column_type {
        ColumnType::Integer => "integer",
        ColumnType::Real => "real",
        ColumnType::String => "string",
        ColumnType::Bitfield => "bitfield",
        ColumnType::Double => "double",
        _ => "ignore",
    }
}

fn main() -> odc::Result<()> {
    let paths: Vec<String> = std::env::args().skip(1).collect();
    if paths.is_empty() {
        eprintln!("Usage:");
        eprintln!("    odc_header <odb2 file 1> [<odb2 file 2> ...]");
        std::process::exit(1);
    }

    // Iterate over all frames in the stream in non-aggregated mode.
    let options = ReaderOptions {
        aggregated: false,
        row_limit: None,
    };

    for path in &paths {
        println!("File: {path}");

        let reader = Reader::from_path_with(path, &options)?;
        for (i, frame) in reader.frames().enumerate() {
            let frame = frame?;

            println!(
                "  Frame: {}, Row count: {}, Column count: {}",
                i + 1,
                frame.row_count(),
                frame.column_count()
            );

            for (key, value) in frame.properties() {
                println!("  Property: {key} => {value}");
            }

            for (col, column) in frame.columns().iter().enumerate() {
                println!(
                    "    Column: {}, Name: {}, Type: {}, Size: {}",
                    col + 1,
                    column.name,
                    type_name(column.column_type),
                    column.decoded_size
                );

                for (bf, bit) in column.bitfield.iter().enumerate() {
                    println!(
                        "      Bitfield: {}, Name: {}, Offset: {}, Nbits: {}",
                        bf + 1,
                        bit.name,
                        bit.offset,
                        bit.size
                    );
                }
            }

            println!();
        }
    }

    Ok(())
}
